### PR TITLE
Implement Runtime Private Environments

### PR DESCRIPTION
**Pull request by raskad**
This Pull Request changes the following:

- Implement runtime private environments and private names conforming with the spec.
- Remove parse-time distinction and mutation of private names.
- Adjust early error handling for private names to conform with the spec.

### CODE REVIEW

File: boa_engine/src/vm/code_block.rs [hunk 1]
```diff
@@ -384,6 +384,11 @@ impl CodeBlock {
                     interner.resolve_expect(self.private_names[operand as usize].description()),
                 )
             }
+            Opcode::PushPrivateEnvironment => {
+                let count = self.read::<u32>(*pc);
+                *pc += size_of::<u32>() * (count as usize + 1);
+                String::new()
```
Comment (HalidOdat) [inline boa_engine/src/vm/code_block.rs:390]: It would be nice to print the operands

**Review by HalidOdat — COMMENTED** [review]
Overall it looks good! Just had a small concern about the `id` field. The id (pointer address) is valid only for how long the Object is (because it can be reassigned to another object).

Can it be used after the object is deallocated?

**Comment by raskad** [comment]
> Overall it looks good! Just had a small concern about the `id` field. The id (pointer address) is valid only for how long the Object is (because it can be reassigned to another object).
> 
> Can it be used after the object is deallocated?

I thought a bit about this. I think this should work. The private environments are only stored on function objects that are methods of classes. Those function objects always have a reference back to the class object with the `class_object` field (I think in theory the class object should be in the environment but that leads to the same outcome). So as long as the function objects with the private environments are alive, the class objects with the `id` pointer must also be alive.

**Review by HalidOdat — APPROVED** [review]
Looks perfect to me!

**Review by jedel1043 — APPROVED** [review]
Great work!

**Pull request merged**